homeassistant/components/tplink/switch.py

```diff
@@ -38,14 +39,14 @@
     )
 
 
-def add_entity(device: SmartPlug, async_add_entities):
+def add_entity(hass: HomeAssistantType, device: SmartPlug, async_add_entities):
     """Check if device is online and add the entity."""
     # Attempt to get the sysinfo. If it fails, it will raise an
     # exception that is caught by async_add_entities_retry which
     # will try again later.
     device.get_sysinfo()
```

Comment:
Does this do I/O?

Reply:
get_sysinfo does not perform I/O. Does that mean add_entity should be async?

Reply:
Actually the get_sysinfo() does nothing but print the already discovered attributes. Considering that the pyHS100 Discover method seems to already bring back the attributes and store it in the get_sysinfo...it seems we should be parsing that (which is the point of the method) instead of doing an update_before_add. I would think this would prevent the issues of the devices sometimes not showing up with the proper name. I am going to test that out (if nothing else, if the value is not None it seems we should use it instead of querying the device again for it's mac, alias, model and device_id since those shouldn't change mid add).

Reply:
We're calling `async_add_entities` from sync context. That's not allowed. We have to change something.

If `device.get_sysinfo` is safe to call within the event loop we can change this callback to run in the event loop by decorating it with `@callback` imported from core.py. Also rename it to `async_add_entity`.

Reply:
If the info isn't cached yet, it will do I/O:
https://github.com/GadgetReactor/pyHS100/blob/59424d2738866a2bf88f8e256ece2baa06f0b024/pyHS100/smartdevice.py#L153

Reply:
Understood, that makes sense about the potential I/O.

Based on the possibility of needing to do I/O from get_sysinfo(), shoud we be using the non async versions of add_entities and setup_entry? I assume that since we are relying on polling that those would always be sync functions. 

While the current async_add_entities_retry does seem to ensure that all devices are added, the fact that it is relying on the "update" call means it adds devices with no name (or as unnamed_device).
 
When I run the Discover.discover(), it seems to me that it always discovers all of the devices (currently I am at 23 devices a mix of static and dhcp ips), so I just assumed relying on the get_sysinfo data was safe (but I get that is not the case in terms if making I/O).

Reply:
We can schedule `async_add_entities` on the event loop from inside sync context with `hass.add_job`.

Reply:
Currently add_entry is called from common using hass.add_job.

After some more testing I updated the device init to set the device id and alias immediately which prevents the issue of the devices not always showing up with the correct id. When relying on them being "updated", sometimes (most times), you get a few devices (especially on the HS300) show up as unnamed and it would take a few restarts for them to show up properly. I believe this last change is preventing the device from getting hit an extra time (I believe the startup is also quicker now). 

My last commit doesn't affect what was there already as far as async_setup_entry calling another async function, which schedules async_add_entities from a sync function  via hass.add_job. Since get_sysinfo is already called before _init_ it means no additional I/O should be done. In all my test before this last change, I still experienced the "unnamed" issue.

The only issue I still see is due to the asyncio.sleep I still see a 1 second (or amount of SLEEP) delay before updating the status. I am going to do some more test to see if the sleep / delay is actually needed. I know believe the issue before was that if it did fail to connect it was immediately being set to  "self._available = False". I changed it to only set that after MAX_ATTEMPTS has been reached which prevents the device from showing up as unavailable.

I am just confused why am I seeing the SLEEP delay...I am expecting that the first iteration would update immediately. I am going to continue reading through the docs a bit more but hopefully I am on the right track at least. 

btw, thank you very much Martin for all of the help and direction. I have learned a lot these past few days, I really appreciate it!

Reply:
We're in sync context here. `async_add_entities` must be run in async context. Please schedule it with `hass.add_job`.

Reply:
An even better way would be to switch context here:
```py
async def async_add_entity(...):
    await hass.async_add_executor_job(
        device.get_sysinfo
    )

    async_add_entities(...)
```

Reply:
Got it, that makes sense. Will test that.

Reply:
In common.py line 179, should we also change 
```
result = await hass.async_add_job(
                    callback, add_object, add_entities
                )
```

to

```
result = await hass.async_create_task(
                    callback, add_object, async_add_entities
                )
```

Reply:
I think we can await it directly, ie we don't need to use `hass.async_create_task`.